agent/consul/util.go

```diff
@@ -273,89 +273,163 @@ func runtimeStats() map[string]string {
 	}
 }
 
-// ServersMeetMinimumVersion returns whether the given alive servers are at least on the
-// given Consul version
-func ServersMeetMinimumVersion(members []serf.Member, minVersion *version.Version) bool {
-	return ServersMeetRequirements(members, func(srv *metadata.Server) bool {
-		return srv.Status != serf.StatusAlive || !srv.Build.LessThan(minVersion)
-	})
+// checkServersProvider exists so that we can unit tests the requirements checking functions
+// without having to spin up a whole agent/server.
+type checkServersProvider interface {
+	CheckServers(datacenter string, fn func(*metadata.Server) bool)
+}
+
+// serverRequirementsFn should inspect the given metadata.Server struct
+// and return two booleans. The first indicates whether the given requirements
+// are met. The second indicates whether this server should be considered filtered.
+//
+// The reason for the two booleans is so that a requirement function could "filter"
+// out the left server members if we only want to consider things which are still
+// around or likely to come back (failed state).
+type serverRequirementFn func(*metadata.Server) (ok bool, filtered bool)
+
+type serversMeetRequirementsState struct {
+	// meetsRequirements is the callback to actual check for some specific requirement
+	meetsRequirements serverRequirementFn
+
+	// ok indicates whether all unfiltered servers meet the desired requirements
+	ok bool
+
+	// found is a boolean indicating that the meetsRequirement function did not filter
+	// at least one of the servers it was asked to process
+	found bool
+}
+
+func (s *serversMeetRequirementsState) update(srv *metadata.Server) bool {
+	ok, filtered := s.meetsRequirements(srv)
+
+	if filtered {
+		// keep going but don't update any of the internal state as this server
+		// was filtered by the requirements function
+		return true
+	}
+
+	// mark that at least one server processed was not filtered
+	s.found = true
+
+	if !ok {
+		// mark that at least one server does not meet the requirements
+		s.ok = false
+
+		// prevent continuing server evaluation
+		return false
+	}
+
+	// this should already be set but this will prevent accidentally reusing
+	// the state object from causing false-negatives.
+	s.ok = true
+
+	// continue evaluating servers
+	return true
+}
+
+// ServersMeetRequirements returns whether the given server members meet the requirements as defined by the
+// callback function and whether any servers were not filtered in the requirements function.
```

Comment:
> and whether any servers were not filtered in the requirements function.

I thought this was wrong for a second for same grammatical reason as the comment above 😄. Does this make it clearer?

```suggestion
// callback function and whether at least one server remains unfiltered by the requirements function.
```

Reply:
I like your wording better.